<a href="https://colab.research.google.com/github/dikiwahyudi11/latihan1_kompstat/blob/main/Latihan_4_Komputasi_Statistika.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center><h1><strong><font color="blue">Studi Kasus Data Wrangling</font></strong></h1></center>

![taudata Analytics](https://github.com/taudataid/eLearning/raw/master/images/covers/cover_ui-n-taudata.jpg "taudata Analytics")

# **LATIHAN 4 - Data Wrangling**
## **Anggota Kelompok :**
### 1. Muhamad Ido Raskapati (1906354154)
### 2. Hanifah Sulasri (1906376810)
### 3. Rafi Alvanzah (2006528736)
### 4. Diki Wahyudi (2106709131)
### 5. Audrey Clarissa (2106724864)
### 6. Rachel Thyffani Margaretha S (2106726900)

In [1]:
import warnings; warnings.simplefilter('ignore')
import pandas as pd

try: # Running Locally, yakinkan "file_" berada di folder "data"
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')
except: # Running in Google Colab
    !mkdir data
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_nilai.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/mengajar.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_dosen.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_kelas.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_krs.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mahasiswa.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mata_kuliah.csv
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')

mengajar.shape, dosen.shape, kelas.shape, krs.shape, mahasiswa.shape, matkul.shape, nilai.shape

((14, 4), (4, 6), (4, 4), (7, 6), (15, 8), (9, 3), (33, 7))

# No 01
Apakah ada dosen yang cenderung memberikan nilai cenderung lebih tinggi (atau lebih rendah) dibandingkan dosen lainnya?

In [2]:
# Code untuk Soal No 01 diberikan sebagai Contoh dan Bonus
# Pelajari dengan baik

nilai_dosen = nilai.groupby(nilai['NIDN']).mean()[['nilai']]
nilai_dosen = nilai_dosen.rename({'nilai': 'mean_nilai'}, axis=1) 
nilai_dosen = pd.merge(dosen, nilai_dosen, left_on='NIDN', right_index=True)
nilai_dosen = nilai_dosen[['NIDN', 'nama', 'mean_nilai']]
sb = nilai.groupby(nilai['NIDN']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1) 
nilai_dosen = nilai_dosen.set_index(['NIDN']) 
nilai_dosen = pd.concat([nilai_dosen, sb], axis=1)
nilai_dosen

nama  mean_nilai  std_nilai
NIDN                                      
123    Andri Basuki   85.750000   7.741355
234        zulfikar   80.428571  15.883504
345   Annisa Fitria   60.090909  16.220077
787            yudi   70.571429  13.010984

### Interpretasi No.1:

Terlihat bahwa terdapat perbedaan pemberian nilai di antara para dosen. 
Dosen yang cenderung memberikan nilai cenderung lebih tinggi adalah **Andri Basuki**, dengan rata rata nilai yang diberikan adalah 85.75. 
Sedangkan Dosen yang cenderung memberikan nilai cenderung lebih rendah adalah **Annisa Fitria**, dengan rata rata nilai yang diberikan adalah 60.09. 


# No 02
Menurut data yang ada, apakah terdapat perbedaan dalam pemberian nilai antara dosen dengan Pendidikan Terakhir S3 dan S2?

In [3]:
dospend = pd.merge(dosen, nilai)
dosen_pend = dospend.groupby(dospend['Pendidikan_Terakhir']).mean()[['nilai']]
dosen_pend = dosen_pend.rename({'nilai':'mean_nilai'},axis=1)
stdscore = dospend.groupby(dospend['Pendidikan_Terakhir']).std()[['nilai']]
stdscore = stdscore.rename({'nilai':'std_nilai'}, axis=1)
dosen_pend = pd.concat([dosen_pend, stdscore], axis=1)
dosen_pend

mean_nilai  std_nilai
Pendidikan_Terakhir                       
S2                    64.166667  15.561075
S3                    83.266667  12.068061

Interpretasi No.2:

Terlihat adanya perbedaan dalam pemberian nilai antara dosen dengan pendidikan terakhir S3 dan S2. Dosen dengan pendidikan terakhir S3 cenderung memiliki rata-rata pemberian nilai yang lebih tinggi, yaitu sebesar 83.27, jika dibandingkan dengan dosen den pendidikan terakhir S2 dengan rata-rata pemberian nilai sebesar 64.17.

# No 03
Apakah terdapat perbedaan performa akademik antara mahasiswa yang berasal dari Madrasah/SMK dan SMU?

In [4]:
mahasiswapend = pd.merge(mahasiswa,nilai)
mahasiswa_pend = mahasiswapend.groupby(mahasiswapend['Asal_Sekolah']).mean()[['nilai']]
mahasiswa_pend = mahasiswa_pend.rename({'nilai':'mean_nilai'},axis=1)
stdscore = mahasiswapend.groupby(mahasiswapend['Asal_Sekolah']).std()[['nilai']]
stdscore = stdscore.rename({'nilai':'std_nilai'}, axis=1)
mahasiswa_pend = pd.concat([mahasiswa_pend, stdscore], axis=1)
mahasiswa_pend

mean_nilai  std_nilai
Asal_Sekolah                       
MA             70.777778  21.111477
SMK            82.818182  11.771306
SMU            65.846154  14.200036

Interpretasi No.3:

Terlihat perbedaan dalam performa akademik antara mahasiswa yang berasal dari MA, SMK, dan SMU. Mahasiswa dengan pendidikan terakhir MA memiliki nilai rata-rata 70.78, sedangkan mahasiswa dengan pendidikan terakhir SMK memiliki nilai rata-rata 82.82, dan terakhir mahasiswa dengan pendidikan terakhir SMU memiliki nilai rata-rata 65.85.

# No 04
Apakah ada perbedaan yang nampak di data terhadap kemampuan akademik antara mahasiswa dan mahasiswi?

In [5]:
# Nomor 4.1
data_mean = nilai.groupby(["nim"]).mean()[["nilai"]]
data_mean = pd.merge(data_mean, mahasiswa, on = 'nim')
data_mean
mean = data_mean.groupby(['Gender']).mean()[['nilai']]
mean = mean.rename({'nilai': 'mean_nilai'}, axis = 1) 
sd = data_mean.groupby(['Gender']).std()[['nilai']]
sd = sd.rename({'nilai': 'std_nilai'}, axis = 1) 
summary  = pd.concat([mean, sd], axis = 1)
summary

mean_nilai  std_nilai
Gender                       
Pria     75.750000  12.762249
Wanita   72.333333  16.132782

In [6]:
# Nomor 4.2
datam = pd.merge(mahasiswa,nilai)
meanm = datam.groupby(datam['Gender']).mean()[['nilai']]
meanm = meanm.rename({'nilai':'mean_nilai'},axis=1)
sdm = datam.groupby(datam['Gender']).std()[['nilai']]
sdm = sdm.rename({'nilai':'std_nilai'}, axis=1)
summary2 = pd.concat([meanm, sdm], axis=1)
summary2

mean_nilai  std_nilai
Gender                       
Pria     73.888889  17.146047
Wanita   71.600000  17.112235


- Interpretasi Nomor 4:

Dari output tersebut, terlihat bahwa rata-rata nilai dari pria dan wanita tidak terlalu jauh berbeda. Selain itu, standar deviasinya juga tidak terlalu jauh berbeda. Oleh karena itu, dapat disimpulkan bahwa tidak terdapat perbedaan yang signifkan di data terhadap kemampuan akademik antara mahasiswa dan mahasiswi.

In [7]:
mahasiswa

nim            nama                    email  Gender Asal_Sekolah  \
0   194015   Muhammad Amir       Samarinda@inet.com    Pria          SMU   
1   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
2   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
3   194003            Udin            Udin@inet.com    Pria           MA   
4   194004            Iwan            Iwan@inet.com    Pria           MA   
5   194005  Bambang Arifin  Bambang Arifin@inet.com    Pria          SMK   
6   194006            Wati            Wati@inet.com  Wanita          SMK   
7   194007             Ani             Ani@inet.com  Wanita          SMK   
8   194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
9   194009            Asep            Asep@inet.com    Pria          SMU   
10  194010            Rina            rina@inet.com  Wanita           MA   
11  194011            Sari            sari@inet.com  Wanita           MA   
12  194012            Bowo            Bowo@inet.com    Pria          SMU   
13  194013            Budi            Budi@inet.com    Pria          SMK   
14  194014           Putri           putri@inet.com  Wanita          SMK   

    Telephone  Angkatan  Asal_Daerah  
0    12345678      2019  Luar Daerah  
1    12345679      2019        Lokal  
2    12345680      2019  Luar Daerah  
3    12345681      2019        Lokal  
4    12345682      2019  Luar Daerah  
5    12345683      2019        Lokal  
6    12345684      2019        Lokal  
7    12345685      2019  Luar Daerah  
8    12345686      2020        Lokal  
9    12345687      2020  Luar Daerah  
10   12345688      2020        Lokal  
11   12345689      2020        Lokal  
12   12345690      2020  Luar Daerah  
13   12345691      2020        Lokal  
14   12345692      2020        Lokal

# No 05
Apakah ada perbedaan prestasi akademik yang nampak di data antar angkatan mahasiswa?

In [8]:
# Cara 4.2 dan 5 sama yaitu rata-rata kan dulu per NIM, bukan di-merge dulu kemudian dirata-ratakan
nilai_mahasiswa = nilai.groupby(nilai['nim']).mean()[['nilai']]
nilai_mahasiswa = nilai_mahasiswa.rename({'nilai': 'mean_nilai'}, axis=1) 
nilai_mahasiswa = pd.merge(mahasiswa, nilai_mahasiswa, left_on='nim', right_index=True)
nilai_mahasiswa = nilai_mahasiswa[['nim', 'Angkatan', 'mean_nilai']]
sb = nilai.groupby(nilai['nim']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1) 
nilai_mahasiswa = nilai_mahasiswa.set_index(['nim']) 
nilai_mahasiswa = pd.concat([nilai_mahasiswa, sb], axis=1)

nilai_Angkatan = nilai_mahasiswa.groupby(nilai_mahasiswa['Angkatan']).mean()[['mean_nilai']]

sb = nilai_mahasiswa.groupby(nilai_mahasiswa['Angkatan']).std()[['mean_nilai']]
sb = sb.rename({'mean_nilai': 'std_nilai'}, axis=1) 
sb

nilai_AngkatanStd = pd.merge(nilai_Angkatan, sb, left_on='Angkatan', right_index=True)
nilai_AngkatanStd

mean_nilai  std_nilai
Angkatan                       
2019        75.59375  13.433658
2020        73.00000  15.044379

Interpretasi 05:

Variabel nilai_mahasiswa untuk menggabungkan nim dan nilai, lalu dihitung mean_nilai dan std_nilai setiap mahasiswa, selanjutnya groupby Angkatan, dan didapatkan hasil seperti di atas.
Mean Nilai Angkatan 2019 dan 2020 tidak berbeda jauh, Standard deviasi masing-masing angkatan cukup besar, artinya masing-masing angkatan memiliki banyak variasi nilai. Dapat disimpulkan kedua angkatan tersebut tidak terdapat perbedaan prestasi akademik.

# No 06
Apakah ada permasalahan di data atau informasi lain yang anda temukan di data?

Jawab:
**Ada**, berikut beberapa permasalahan pada data:
1. Terdapat kolom ```rand``` pada data ```nilai```


In [9]:
nilai.head()
nilai.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33 entries, 0 to 32
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   nim             33 non-null     int64  
 1   id_mata_kuliah  33 non-null     object 
 2   NIDN            33 non-null     int64  
 3   nilai           33 non-null     int64  
 4   Semester        33 non-null     object 
 5   Tahun           33 non-null     int64  
 6   rand            1 non-null      float64
dtypes: float64(1), int64(4), object(2)
memory usage: 1.9+ KB


Terlihat bahwa kolom ```rand``` hanya memiliki 1 entry di dalamnya. maka akan dihilangkan kolom tersebut dari dataframe

In [10]:
#kode untuk menghilangkan column rand
nilai = nilai.drop('rand', axis=1)
nilai.head()

nim id_mata_kuliah  NIDN  nilai Semester  Tahun
0  194015          KAL01   123     70   ganjil   2019
1  194001          KAL01   123     80   ganjil   2019
2  194002          KAL01   123     85   ganjil   2019
3  194003          KAL01   123     90   ganjil   2019
4  194004          KAL01   123     92   ganjil   2019

2. Terdapat duplicate pada data ```KRS```

In [11]:
krs

id id_mata_kuliah kode_kelas     nim  tahun semester
0   1          KAL01      SIS1J  193001   2021    genap
1   2          STA01       D101  193001   2021    genap
2   3           OR01       D212  193001   2021    genap
3   4          MK002       P601  194001   2021    genap
4   5          MK002       P601  194001   2021    genap
5   6          MK003      SIS1J  194001   2021    genap
6   7          MK004       D101  194001   2021    genap

In [12]:
#kode untuk menghilangkan duplicate di KRS
krs = krs.drop_duplicates(['id_mata_kuliah', 'kode_kelas', 'nim'])
krs

id id_mata_kuliah kode_kelas     nim  tahun semester
0   1          KAL01      SIS1J  193001   2021    genap
1   2          STA01       D101  193001   2021    genap
2   3           OR01       D212  193001   2021    genap
3   4          MK002       P601  194001   2021    genap
5   6          MK003      SIS1J  194001   2021    genap
6   7          MK004       D101  194001   2021    genap

3. Data pada tabel mahasiswa tidak urut berdasarkan NIM, akan dilakukan sorting

In [13]:
mahasiswa

nim            nama                    email  Gender Asal_Sekolah  \
0   194015   Muhammad Amir       Samarinda@inet.com    Pria          SMU   
1   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
2   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
3   194003            Udin            Udin@inet.com    Pria           MA   
4   194004            Iwan            Iwan@inet.com    Pria           MA   
5   194005  Bambang Arifin  Bambang Arifin@inet.com    Pria          SMK   
6   194006            Wati            Wati@inet.com  Wanita          SMK   
7   194007             Ani             Ani@inet.com  Wanita          SMK   
8   194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
9   194009            Asep            Asep@inet.com    Pria          SMU   
10  194010            Rina            rina@inet.com  Wanita           MA   
11  194011            Sari            sari@inet.com  Wanita           MA   
12  194012            Bowo            Bowo@inet.com    Pria          SMU   
13  194013            Budi            Budi@inet.com    Pria          SMK   
14  194014           Putri           putri@inet.com  Wanita          SMK   

    Telephone  Angkatan  Asal_Daerah  
0    12345678      2019  Luar Daerah  
1    12345679      2019        Lokal  
2    12345680      2019  Luar Daerah  
3    12345681      2019        Lokal  
4    12345682      2019  Luar Daerah  
5    12345683      2019        Lokal  
6    12345684      2019        Lokal  
7    12345685      2019  Luar Daerah  
8    12345686      2020        Lokal  
9    12345687      2020  Luar Daerah  
10   12345688      2020        Lokal  
11   12345689      2020        Lokal  
12   12345690      2020  Luar Daerah  
13   12345691      2020        Lokal  
14   12345692      2020        Lokal

In [14]:
#kode untuk urut mahasiswa by NIM
mahasiswa = mahasiswa.sort_values(by=['nim'])
mahasiswa

nim            nama                    email  Gender Asal_Sekolah  \
1   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
2   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
3   194003            Udin            Udin@inet.com    Pria           MA   
4   194004            Iwan            Iwan@inet.com    Pria           MA   
5   194005  Bambang Arifin  Bambang Arifin@inet.com    Pria          SMK   
6   194006            Wati            Wati@inet.com  Wanita          SMK   
7   194007             Ani             Ani@inet.com  Wanita          SMK   
8   194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
9   194009            Asep            Asep@inet.com    Pria          SMU   
10  194010            Rina            rina@inet.com  Wanita           MA   
11  194011            Sari            sari@inet.com  Wanita           MA   
12  194012            Bowo            Bowo@inet.com    Pria          SMU   
13  194013            Budi            Budi@inet.com    Pria          SMK   
14  194014           Putri           putri@inet.com  Wanita          SMK   
0   194015   Muhammad Amir       Samarinda@inet.com    Pria          SMU   

    Telephone  Angkatan  Asal_Daerah  
1    12345679      2019        Lokal  
2    12345680      2019  Luar Daerah  
3    12345681      2019        Lokal  
4    12345682      2019  Luar Daerah  
5    12345683      2019        Lokal  
6    12345684      2019        Lokal  
7    12345685      2019  Luar Daerah  
8    12345686      2020        Lokal  
9    12345687      2020  Luar Daerah  
10   12345688      2020        Lokal  
11   12345689      2020        Lokal  
12   12345690      2020  Luar Daerah  
13   12345691      2020        Lokal  
14   12345692      2020        Lokal  
0    12345678      2019  Luar Daerah

4. Terdapat informasi yang kurang lengkap pada dataset setelah dilakukan merging

In [15]:
#kode untuk merging seluruh dataset
a = pd.merge(mengajar, dosen, on='NIDN', how='outer')
b = pd.merge(krs, kelas, on='kode_kelas', how='outer')
c = pd.merge(mahasiswa, b, on='nim', how='outer')
d = pd.merge(nilai, matkul, on='id_mata_kuliah', how='outer')
e = pd.merge(d, c, on=['nim', 'id_mata_kuliah'], how='outer')
f = pd.merge(e, a, on=['NIDN', 'Semester', 'Tahun'], how='outer', suffixes=('','_dosen'))

#jumlah nan pada setiap kolom 
f.isna().sum()

nim                    15
id_mata_kuliah         24
NIDN                   25
nilai                  35
Semester               25
Tahun                  25
nama_mata_kuliah       30
SKS                    30
nama                   59
email                  59
Gender                 59
Asal_Sekolah           59
Telephone              59
Angkatan               59
Asal_Daerah            59
id                     70
kode_kelas             70
tahun                  70
semester               70
nama_kelas             70
lokasi_kelas           70
Kapasitas              70
kode_MK                39
nama_dosen             39
email_dosen            39
Pendidikan_Terakhir    39
Telephone_dosen        39
Jenis_Kelamin          39
dtype: int64

5. Melanjutkan nomor 4, salah satu penyebabnya adalah data KRS tidak lengkap

In [16]:
krs_mahasiswa=pd.merge(mahasiswa, krs, on='nim', how ='left')
krs_mahasiswa

nim            nama                    email  Gender Asal_Sekolah  \
0   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
1   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
2   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
3   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
4   194003            Udin            Udin@inet.com    Pria           MA   
5   194004            Iwan            Iwan@inet.com    Pria           MA   
6   194005  Bambang Arifin  Bambang Arifin@inet.com    Pria          SMK   
7   194006            Wati            Wati@inet.com  Wanita          SMK   
8   194007             Ani             Ani@inet.com  Wanita          SMK   
9   194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
10  194009            Asep            Asep@inet.com    Pria          SMU   
11  194010            Rina            rina@inet.com  Wanita           MA   
12  194011            Sari            sari@inet.com  Wanita           MA   
13  194012            Bowo            Bowo@inet.com    Pria          SMU   
14  194013            Budi            Budi@inet.com    Pria          SMK   
15  194014           Putri           putri@inet.com  Wanita          SMK   
16  194015   Muhammad Amir       Samarinda@inet.com    Pria          SMU   

    Telephone  Angkatan  Asal_Daerah   id id_mata_kuliah kode_kelas   tahun  \
0    12345679      2019        Lokal  4.0          MK002       P601  2021.0   
1    12345679      2019        Lokal  6.0          MK003      SIS1J  2021.0   
2    12345679      2019        Lokal  7.0          MK004       D101  2021.0   
3    12345680      2019  Luar Daerah  NaN            NaN        NaN     NaN   
4    12345681      2019        Lokal  NaN            NaN        NaN     NaN   
5    12345682      2019  Luar Daerah  NaN            NaN        NaN     NaN   
6    12345683      2019        Lokal  NaN            NaN        NaN     NaN   
7    12345684      2019        Lokal  NaN            NaN        NaN     NaN   
8    12345685      2019  Luar Daerah  NaN            NaN        NaN     NaN   
9    12345686      2020        Lokal  NaN            NaN        NaN     NaN   
10   12345687      2020  Luar Daerah  NaN            NaN        NaN     NaN   
11   12345688      2020        Lokal  NaN            NaN        NaN     NaN   
12   12345689      2020        Lokal  NaN            NaN        NaN     NaN   
13   12345690      2020  Luar Daerah  NaN            NaN        NaN     NaN   
14   12345691      2020        Lokal  NaN            NaN        NaN     NaN   
15   12345692      2020        Lokal  NaN            NaN        NaN     NaN   
16   12345678      2019  Luar Daerah  NaN            NaN        NaN     NaN   

   semester  
0     genap  
1     genap  
2     genap  
3       NaN  
4       NaN  
5       NaN  
6       NaN  
7       NaN  
8       NaN  
9       NaN  
10      NaN  
11      NaN  
12      NaN  
13      NaN  
14      NaN  
15      NaN  
16      NaN

Jika dimerge, terlihat bahwa banyak mahasiswa yang mata kuliahnya kosong, sehingga data akan dibersihkan menggunakan ```.dropna()```

In [17]:
krs_mahasiswa.dropna()

nim           nama              email Gender Asal_Sekolah  Telephone  \
0  194001  Samsul Kurnia  Bontang@email.com   Pria          SMU   12345679   
1  194001  Samsul Kurnia  Bontang@email.com   Pria          SMU   12345679   
2  194001  Samsul Kurnia  Bontang@email.com   Pria          SMU   12345679   

   Angkatan Asal_Daerah   id id_mata_kuliah kode_kelas   tahun semester  
0      2019       Lokal  4.0          MK002       P601  2021.0    genap  
1      2019       Lokal  6.0          MK003      SIS1J  2021.0    genap  
2      2019       Lokal  7.0          MK004       D101  2021.0    genap

Terlihat bahwa hanya **Samsul Kurnia** yang memiliki data KRS pada semester genap 2021

Jawab:
**Ada**, berikut beberapa informasi tambahan pada data:

1. Tingkat kesulitan mata kuliah

In [18]:
u = pd.merge(nilai, mahasiswa, on='nim')
u = u.groupby(['id_mata_kuliah'])[['nilai']].mean()
u = u.rename({'nilai':'rata-rata nilai'}, axis=1)
u

rata-rata nilai
id_mata_kuliah                 
CS01                  80.428571
KAL01                 78.666667
KAL02                 63.500000
STA02                 58.142857

Interpretasi Informasi 1:

Dapat dilihat bahwa rata-rata nilai CS01 memiliki nilai paling tinggi ini dapat menginplikasikan bahwa CS01 merupakan matakul paling mudah diantara ke-empat matkul lainnya

2. Penurunan nilai mata kuliah pertahun

In [19]:
y = nilai.groupby(['id_mata_kuliah', 'Tahun'])[['nilai']].mean()
y

nilai
id_mata_kuliah Tahun           
CS01           2020   80.428571
KAL01          2019   85.750000
               2020   70.571429
KAL02          2020   63.500000
STA02          2021   58.142857

Interpretasi Informasi 2:

Dapat dilihat bahwa hanya matkul KAL01 yang telah dilaksanakan selama 2 semester dan rata-rata nilai KAL01 mengalami penurunan dari tahun sebelumnya.